### ISSUE TITLE
Provide some way to transmit LS crashes with info via telemetry

### ISSUE THREAD

**Issue by davidanthoff** [Microsoft/vscode-languageserver-node#296]
We know our language server (julia language) process crashes from time to time because of bugs in our code. In that case the LS server process will print a stack trace to STDERR and then the process will shutdown.

The content of the julia STDERR stream normally appears in the julia language server output channel, which is super handy. I kind of assume that this https://github.com/Microsoft/vscode-languageserver-node/blob/master/client/src/client.ts#L2505 line does that?

In any case, it would be super useful if we could somehow transmit what came in via STDERR from the julia language server process via telemetry to us if the julia language server process crashes. I kind of have a vague sense how I might be able to hook into an event when our LS process crashes, but it is not clear to me how I might be able to access what came in from its STDERR. Probably I would somehow have to hook into the point where this gets written to the output channel, and then also write it into some kind of buffer, that we could then transmit via telemetry whenever we detect a LS crash?

I think this kind of feature would be the single most helpful things for us to improve the stability and quality of our extension.

**Comment by rcjsuen** [comment]
You might also be interested Microsoft/language-server-protocol#350.

**Comment by dbaeumer** [comment]
No, this code line you pointed to is actually not printing this. I the done by the main code which starts the process and then prints stderr to the console. See the code around here: https://github.com/Microsoft/vscode-languageserver-node/blob/master/client\src\main.ts#L273

Depending on the transport kind it prints stdout and stderr or only stderr. Today you can already tweak this to pass a function as the server options and start the server by yourself, however might be cumbersome. So I suggest that we pipe the stderr / out through the middleware as well so that a client can tweak what happens to it. Would you be willing to work on a PR for this?

**Comment by davidanthoff** [comment]
That sounds like a good plan. I don't have the time to work on a PR for that, I'm afraid...

**Comment by tsmaeder** [comment]
@dbaeumer  So you're suggesting sending all stdout to the client? This would include the whole LSP traffic if using stdout as a comms channel, though. 
Would it make sense to just do the equivalent of 'tail -n 500' when the server crashes?

**Comment by dbaeumer** [comment]
Actually the current version of the client is already doing the right thing depending on the transport used. The sdtout / stderr is printed to the output channel.So the only thing necessary would be to allow extensions to listen on that in case they want to log it into telementry. Should be fairly easy to achieve.

**Comment by MikhailArkhipov** [comment]
Would be useful to have dedicated `sendTelemetryErrorEvent` and `sendTelemetryException` in LSP rather than just `logEvent` as `TelemetryReporter` has.

**Comment by dbaeumer** [comment]
@MikhailArkhipov can you explain why. The problem is that the client will never forward telemetry. All it can do is to fire an event when it received it. Telemetry must always be handled by the extension itself.

**Comment by MikhailArkhipov** [comment]
@dbaeumer - I have heard `sendTelemetryException` sanitizes possible PII data from exceptions.

**Comment by dbaeumer** [comment]
IMO LSP shouldn't be in that business. It is up to the extension to decide this and it highly depends on the telemetry system used.
